src/flow/src/utils.rs

```diff
@@ -167,288 +176,265 @@ impl Arrangement {
-            for (key, updates) in last_batch.1 {
-                let iter = updates
-                    .iter()
-                    .filter(|(_val, ts, _diff)| range.contains(ts))
-                    .map(|(val, ts, diff)| ((key.clone(), val.clone()), *ts, *diff));
-                result.extend(iter);
-            }
-        };
-        result
+        }
+        res
     }
 
-    /// expire keys in now that are older than expire_time, intended for reducing memory usage and limit late data arrive
-    pub fn trunc_expired(&mut self, now: Timestamp) {
+    /// Expire keys in now that are older than expire_time, intended for reducing memory usage and limit late data arrive
+    pub fn truncate_expired_keys(&mut self, now: Timestamp) {
         if let Some(s) = &mut self.expire_state {
-            let expire_time = if let Some(t) = s.compute_expiration_timestamp(now) {
-                t
-            } else {
-                // never expire
-                return;
-            };
-            // find all keys smaller than or equal expire_time and silently remove them
-            let mut after = s.event_ts_to_key.split_off(&(expire_time + 1));
-            std::mem::swap(&mut s.event_ts_to_key, &mut after);
-            let before = after;
-            for key in before.into_iter().flat_map(|i| i.1.into_iter()) {
-                for (_ts, batch) in self.spine.iter_mut() {
-                    batch.remove(&key);
+            if let Some(expired_keys) = s.remove_expired_keys(now) {
+                for key in expired_keys {
+                    for (_, batch) in self.spine.iter_mut() {
+                        batch.remove(&key);
+                    }
                 }
             }
         }
     }
 
-    /// get current state of things
-    /// useful for query existing keys(i.e. reduce and join operator need to query existing state)
-    pub fn get(&self, now: Timestamp, key: &Row) -> Option<(Row, Timestamp, Diff)> {
-        if self.full_arrangement
-            && self
-                .spine
-                .first_key_value()
-                .map(|(ts, _)| *ts >= now)
-                .unwrap_or(false)
+    /// Get current state of things.
+    ///
+    /// Useful for query existing keys (i.e. reduce and join operator need to query existing state)
+    pub fn get(&self, now: Timestamp, key: &Row) -> Option<DiffRow> {
+        // FAST PATH:
+        //
+        // If `now <= lowest_ts in the spine`, and it's full arrangement, we can directly return the value
+        // from the current state (which is the first batch in the spine).
+        if let Some((lowest_ts, batch)) = self.spine.first_key_value()
+            && now <= *lowest_ts
+            && self.full_arrangement
         {
-            self.spine
-                .first_key_value()
-                .and_then(|(_ts, batch)| batch.get(key).and_then(|v| v.first()).cloned())
+            return batch.get(key).and_then(|updates| updates.first().cloned());
+        }
```

Comment:
I've included a test at the bottom (`test_full_arrangement_get_from_first_entry`) to illustrate my confusion. Could you please take a look, @discord9? Is this behavior expected?